clamz.1
.TH clamz 1 "October 2011" "Clamz 0.5"
.SH NAME
clamz \- download MP3 music files from Amazon.com
.SH SYNOPSIS
\fBclamz\fR [ \fIoptions\fR ] \fIamz-file\fR ...

.SH DESCRIPTION
\fBclamz\fR is a little command-line program to download MP3 files
from Amazon.com's music store.  It is intended to serve as a
substitute for Amazon's official MP3 Downloader, which is not free
software (and therefore is only available in binary form for a limited
set of platforms.)  \fBclamz\fR can be used to download either
individual songs or complete albums that you have purchased from
Amazon.

In order to use \fBclamz\fR, you must first enable the ``MP3 dowloader
mode'' by visiting the following URL:
.IP
http://www.amazon.com/gp/dmusic/after_download_manager_install.html?AMDVersion=1.0.9
.PP
If you are outside the US, replace `amazon.com' with the corresponding
domain for your country (e.g., amazon.co.uk for the UK, or amazon.fr
for France.)  Amazon does not allow anyone to purchase MP3 files from
outside their country of residence.

To download the contents of an AMZ file into the current directory,
just run
.IP
clamz \fIsome-file-name\fR.amz
.PP
More advanced options are listed below.

.SS OPTIONS
.TP
\fB-o\fR \fIname-format\fR, \fB--output\fR=\fIname-format\fR
Set the name of the downloaded file(s).  This may contain any of
several variables which are derived from the input AMZ file; see
\fBFORMAT VARIABLES\fR below.  Note that the format string should be
enclosed in single quotes, to stop the shell from expanding variables
itself.
.TP
\fB-d\fR \fIdirectory-format\fR, \fB--output-dir\fR=\fIdirectory-format\fR
Set the directory where the downloaded files should be placed.  (If
this directory does not exist, it will be created.)  This may also
contain format variables.
.TP
\fB-r\fR, \fB--resume\fR
Resume downloading a partially-downloaded album.  (By default, if you
are downloading a file named foo.mp3, and the file foo.mp3 already
exists in the destination directory, \fBclamz\fR will rename the new
file to foo.mp3.1 to avoid overwriting the old file.  If the \fB-r\fR
option is used, \fBclamz\fR will instead assume that the first part of
the file has already been downloaded, and will resume downloading from
where it left off.)
.TP
\fB-i\fR, \fB--info\fR
Rather than downloading anything, just display detailed information
about the given AMZ file(s) to standard output.
.TP
\fB-x\fR, \fB--xml\fR
Rather than downloading anything, print the raw, decrypted XML data
from the AMZ file to standard output.
.TP
\fB-v\fR, \fB--verbose\fR
Display detailed information while downloading.
.TP
\fB-q\fR, \fB--quiet\fR
Turn off the normal progress display; display only error messages.
.TP
\fB--forbid-chars\fR=\fIcharacters\fR
Do not allow the given \fIcharacters\fR to be used in output
filenames.  Note that control characters and slashes may never be used
in filenames.
.TP
\fB--allow-chars\fR=\fIcharacters\fR
Opposite of the above; remove the given \fIcharacters\fR from the set
of disallowed characters.
.TP
\fB--allow-uppercase\fR
Allow uppercase letters in filenames.
.TP
\fB--forbid-uppercase\fR
Do not allow uppercase letters in filenames.
.TP
\fB--utf8-filenames\fR
Use UTF-8 when writing filenames (the default behavior is to use UTF-8
if the system locale says so, otherwise ASCII.)
.TP
\fB--ascii-filenames\fR
Use only ASCII characters in filenames.
.TP
\fB--help\fR
Print out a summary of options.
.TP
\fB--version\fR
Print out version information.

.SS FORMAT VARIABLES
As part of a \fIname-format\fR or \fIdirectory-format\fR option, you
may include references to environment variables (e.g., \fB$HOME\fR) or
to the `xdg-user-dirs' configuration variables (e.g.,
\fB$XDG_MUSIC_DIR\fR or \fB$XDG_DESKTOP_DIR\fR.)

In addition, the following special variables are defined for each track,
based on the information provided in the AMZ file, and subject to the
above configuration options (\fB--forbid-chars\fR,
\fB--forbid-uppercase\fR, etc.)
.TP
\fB${title}\fR, \fB${creator}\fR, \fB${tracknum}\fR, \fB${discnum}\fR, \fB${genre}\fR, \fB${asin}\fR
Title, creator, track number, disc number, genre, and ASIN (Amazon
Standard Identification Number) of each individual track.
.TP
\fB${album}\fR, \fB${album_artist}\fR, \fB${album_asin}\fR
Title, primary artist, and ASIN of the album the track comes from.
(This information is available both for single-track and full-album
downloads.)
.TP
\fB${suffix}\fR
Suffix of the output file (currently only `mp3'.)
.TP
\fB${amz_title}\fR, \fB${amz_creator}\fR, \fB${amz_genre}\fR, \fB${amz_asin}\fR
These variables formerly contained metadata for the AMZ file as a
whole; current AMZ files do not contain this information.  Using these
variables is not recommended.
.PP
Similar to shell variable expansion, you can also use the following
conditional expressions:
.TP
\fB${\fIvar\fB:-\fIstring\fB}\fR
Expands to the value of variable \fIvar\fR if it is defined and
non-empty; otherwise, expands to \fIstring\fR (which may itself
contain variable references.)
.TP
\fB${\fIvar\fB:+\fIstring\fB}\fR
Expands to \fIstring\fR if the variable \fIvar\fR is defined and
non-empty; otherwise, expands to an empty string.
.PP
Note that when you include format variables in a command-line
argument, you will usually need to enclose the argument in single
quotes ('...'), or insert a backslash before the `$', to prevent the
shell from trying to expand the variables itself.

.SH FILES
.TP
$HOME/.clamz/config
General configuration file, providing default settings for the
\fB--output\fR, \fB--forbid-chars\fR, \fB--allow-uppercase\fR and
\fB--utf8-filenames\fR options.
.TP
$HOME/.clamz/amzfiles/
Directory containing backup copies of AMZ files.
.TP
$HOME/.clamz/logs/
Directory containing log files.

.SH ENVIRONMENT
.TP
http_proxy
HTTP proxy to use for downloading files; use `server:port' to specify
a port number.

.SH AUTHOR
Benjamin Moody <floppusmaximus@users.sf.net>


clamz.c
/*
 * clamz - Command-line downloader for the Amazon.com MP3 store
 * Copyright (c) 2008-2010 Benjamin Moody
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <ctype.h>

#include <locale.h>
#include <langinfo.h>

#include <curl/curl.h>
#include <gcrypt.h>

#include "clamz.h"

static void print_pl_info(clamz_playlist *pl, const char* fname)
{
  clamz_meta_list *meta;

  printf("Playlist: %s\n", fname);
  if (pl->title)
    printf("* Title:    %s\n", pl->title);
  if (pl->creator)
    printf("* Creator:  %s\n", pl->creator);
  if (pl->image_name)
    printf("* Image:    %s\n", pl->image_name);
  for (meta = pl->meta; meta; meta = meta->next) {
    if (!strcmp(meta->urn, PMETA_ASIN))
      printf("* ASIN:     %s\n", meta->value);
    else if (!strcmp(meta->urn, PMETA_GENRE))
      printf("* Genre:    %s\n", meta->value);
    else
      printf("* '%s' = %s\n", meta->urn, meta->value);
  }
}

static void print_tr_info(clamz_track *tr, int n)
{
  clamz_meta_list *meta;

  printf("\n  Track %d:\n", n);
  if (tr->location)
    printf("  - URL:           %s\n", tr->location);
  if (tr->title)
    printf("  - Title:         %s\n", tr->title);
  if (tr->creator)
    printf("  - Creator:       %s\n", tr->creator);
  if (tr->album)
    printf("  - Album:         %s\n", tr->album);
  if (tr->image_name)
    printf("  - Image:         %s\n", tr->image_name);
  if (tr->duration)
    printf("  - Duration:      %s\n", tr->duration);
  if (tr->trackNum)
    printf("  - Track Number:  %s\n", tr->trackNum);
  for (meta = tr->meta; meta; meta = meta->next) {
    if (!strcmp(meta->urn, TMETA_ALBUM_ARTIST))
      printf("  - Album Artist:  %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_ALBUM_ASIN))
      printf("  - Album ASIN:    %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_ASIN))
      printf("  - ASIN:          %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_DISC_NUM))
      printf("  - Disc Number:   %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_FILE_SIZE))
      printf("  - File Size:     %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_GENRE))
      printf("  - Genre:         %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_PRODUCT_TYPE))
      printf("  - Product Type:  %s\n", meta->value);
    else if (!strcmp(meta->urn, TMETA_TRACK_TYPE))
      printf("  - File Type:     %s\n", meta->value);
    else
      printf("  - '%s' = %s\n", meta->urn, meta->value);
  }
}

static const char *getbasename(const char *fname)
{
  const char *b;

  while ((b = strchr(fname, '/')))
    fname = b + 1;

  return fname;
}

static int run_amz_file(clamz_downloader *dl,
			const clamz_config *cfg,
			FILE *amzfile,
			const char *fname)
{
  char *inbuf;
  unsigned char *xml;
  size_t sz;
  clamz_playlist *pl;
  int i;
  int status, rv = 0;
  char *logname;
  FILE *logfile;

  sz = 0;
  inbuf = NULL;

  while (!feof(amzfile) && !ferror(amzfile)) {
    if (inbuf)
      inbuf = realloc(inbuf, (sz + 1024) * sizeof(char));
    else
      inbuf = malloc((sz + 1024) * sizeof(char));
    sz += fread(&inbuf[sz], 1, 1024, amzfile);
  }

  if (amzfile != stdin)
    fclose(amzfile);

  if (cfg->printonly && cfg->printasxml) {
    xml = decrypt_amz_file(inbuf, sz, fname);
    if (!xml) {
      free(inbuf);
      return 2;
    }

    printf("%s", xml);
    free(xml);
    free(inbuf);
    return 0;
  }
  else {
    if (!cfg->printonly) {
      if (write_backup_file(inbuf, sz, getbasename(fname))) {
        free(inbuf);
        return 3;
      }
    }

    pl = new_playlist();

    if (read_amz_file(pl, inbuf, sz, fname)) {
      free(inbuf);
      free_playlist(pl);
      return 2;
    }

    if (!cfg->printonly) {
      logname = get_config_file_name("logs", getbasename(fname), ".log");
      if (!logname) {
        free(inbuf);
	return 1;
      }

      logfile = fopen(logname, "w");
      if (!logfile) {
	perror(logname);
	free(logname);
        free(inbuf);
	return 3;
      }

      free(logname);
      set_download_log_file(dl, logfile);
    }
    else {
      logfile = NULL;
    }

    if (cfg->printonly || cfg->verbose)
      print_pl_info(pl, fname);

    for (i = 0; i < pl->num_tracks; i++) {
      if (cfg->printonly || cfg->verbose)
        print_tr_info(pl->tracks[i], i + 1);

      status = download_track(dl, pl->tracks[i]);
      if (!rv)
        rv = status;
      fputc('\n', stderr);
    }

    set_download_log_file(dl, NULL);

    free(inbuf);
    free_playlist(pl);
    if (logfile)
      fclose(logfile);
    return rv;
  }
}

/* Parse XDG user-dirs configuration file and set environment
   variables (XDG_DESKTOP_DIR, XDG_MUSIC_DIR, etc.) */
static void set_xdg_user_dirs()
{
  const char *p, *q, *home;
  char *cfgfname;
  FILE *cfgfile;
  char buf[1024];
  char *envstr;

  home = getenv("HOME");
  if (!home)
    return;

  if ((p = getenv("XDG_CONFIG_HOME"))) {
    cfgfname = strdup(p);
    if (!cfgfname)
      return;
  }
  else {
    cfgfname = strdup(home);
    if (!cfgfname)
      return;
    if (concatenate(&cfgfname, "/.config", strlen("/.config"))) {
      free(cfgfname);
      return;
    }
  }

  if (concatenate(&cfgfname, "/user-dirs.dirs", strlen("/user-dirs.dirs"))) {
    free(cfgfname);
    return;
  }

  cfgfile = fopen(cfgfname, "r");
  free(cfgfname);
  if (!cfgfile)
    return;

  while (fgets(buf, sizeof(buf), cfgfile)) {
    p = buf;
    while (*p == ' ' || *p == '\t')
      p++;
    if (strncmp(p, "XDG_", 4))
      continue;

    q = p;
    while (isalnum((int) (unsigned char) *q) || *q == '_')
      q++;
    if (q == p || q[0] != '=' || q[1] != '"')
      continue;

    q++;
    envstr = NULL;
    if (concatenate(&envstr, p, q - p)) {
      free(envstr);
      fclose(cfgfile);
      return;
    }

    q++;
    if (!strncmp(q, "$HOME", 5)) {
      if (concatenate(&envstr, home, strlen(home))) {
	free(envstr);
	fclose(cfgfile);
	return;
      }
      q += 5;
    }

    while (*q && *q != '"') {
      if (q[0] == '\\' && q[1])
	q++;
      if (concatenate(&envstr, q, 1)) {
	free(envstr);
	fclose(cfgfile);
	return;
      }
      q++;
    }

    putenv(envstr);
  }

  fclose(cfgfile);
}

static int utf8locale;

int main(int argc, char **argv)
{
  clamz_config cfg;
  clamz_downloader *dl;
  char buf[256];
  FILE *amzfile;
  int err = 0;
  int i, n = 0;
  char *str;

  cfg.output_dir = cfg.name_format = cfg.forbid_chars = NULL;
  cfg.allowupper = cfg.allowutf8 = cfg.printonly = cfg.printasxml = 0;
  cfg.verbose = cfg.quiet = cfg.resume = 0;
  cfg.maxattempts = 5;

  /* Disable secure memory; we don't need it. */
  gcry_control(GCRYCTL_DISABLE_SECMEM, 0);
  gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);

  setlocale(LC_ALL, "");
  str = nl_langinfo(CODESET);
  if (!strcasecmp(str, "UTF-8") || !strcasecmp(str, "UTF8"))
    utf8locale = cfg.utf8locale = 1;
  else
    utf8locale = cfg.utf8locale = 0;

  set_xdg_user_dirs();

  if (parse_args(&argc, argv, &cfg)) {
    if (cfg.output_dir) free(cfg.output_dir);
    if (cfg.name_format) free(cfg.name_format);
    if (cfg.forbid_chars) free(cfg.forbid_chars);
    return 1;
  }

  if (!cfg.printonly)
    curl_global_init(CURL_GLOBAL_ALL);

  dl = new_downloader(&cfg);
  if (!dl)
    return 1;

  for (i = 1; i < argc; i++) {
    if (!strcmp(argv[i], "-")) {
      sprintf(buf, "clamz-stdin-%d", getpid());
      err = run_amz_file(dl, &cfg, stdin, buf);
    }
    else {
      amzfile = fopen(argv[i], "rb");
      if (!amzfile) {
	perror(argv[i]);
	err = 2;
	break;
      }
      err = run_amz_file(dl, &cfg, amzfile, argv[i]);
    }

    if (err)
      break;
    else
      n++;
  }

  free_downloader(dl);

  if (!cfg.printonly)
    curl_global_cleanup();

  if (cfg.output_dir) free(cfg.output_dir);
  if (cfg.name_format) free(cfg.name_format);
  if (cfg.forbid_chars) free(cfg.forbid_chars);

  if (!cfg.quiet && !cfg.printonly)
    fprintf(stderr, "%d of %d AMZ files downloaded successfully.\n",
	    n, argc - 1);

  return err;
}


void print_error(const char *message, ...)
{
  static char buf[4096];
  char *p, *q, *r;
  va_list ap;

  fprintf(stderr, "\rERROR: ");

  va_start(ap, message);
  vsnprintf(buf, 4095, message, ap);
  buf[4095] = 0;
  va_end(ap);

  p = q = buf;
  r = NULL;
  while (*p) {
    if (*p == '\n') {
      *p = 0;
      fputs(q, stderr);
      fputc('\n', stderr);
      fputc('\n', stderr);
      p++;
      q = p;
      r = NULL;
    }
    else if (p - q > 70 && r) {
      *r = 0;
      fputs(q, stderr);
      fputc('\n', stderr);
      p = q = r + 1;
      r = NULL;
    }
    else {
      if (*p == ' ')
	r = p;
      p++;
    }
  }
  if (*q) {
    fputs(q, stderr);
    fputc('\n', stderr);
  }

  fputc('\n', stderr);
}

void print_progress(const clamz_track *tr, const char *filename,
		    int progress)
{
  int i, j;
  const char *name;

  fputc('\r', stderr);

  if (tr->title)
    name = tr->title;
  else
    name = filename;

  for (i = j = 0; i < 32 && name[j]; i++) {
    if ((unsigned char) name[j] & 0x80) {
      if (!utf8locale)
	fputc('?', stderr);

      do {
	if (utf8locale)
	  fputc(name[j], stderr);
	j++;
      } while (((unsigned char) name[j] & 0xc0) == 0x80);
    }
    else {
      fputc(name[j], stderr);
      j++;
    }
  }

  for (; i < 32; i++)
    fputc(' ', stderr);

  if (progress >= 0) {
    fputs("  [", stderr);
    for (i = 0; i < (progress / 3); i++)
      fputc('#', stderr);
    for (; i < 33; i++)
      fputc(' ', stderr);
    fprintf(stderr, "]  %2d%% ", progress);
  }
  else {
    fprintf(stderr, "  ...");
  }

  fflush(stderr);
}
